halo2_proofs/src/poly/kzg/multiopen/shplonk.rs

```diff
@@ -0,0 +1,272 @@
+
+impl<F: FieldExt, T: PartialEq + Clone> Commitment<F, T> {
+    fn get(&self) -> T {
+        self.0 .0.clone()
+    }
+
+    fn evals(&self) -> Vec<F> {
+        self.0 .1.clone()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct RotationSet<F: FieldExt, T: PartialEq + Clone> {
+    commitments: Vec<Commitment<F, T>>,
+    points: Vec<F>,
+}
+
+#[derive(Debug, PartialEq)]
+struct IntermediateSets<F: FieldExt, Q: Query<F>> {
+    rotation_sets: Vec<RotationSet<F, Q::Commitment>>,
+    super_point_set: Vec<F>,
+}
+
+fn construct_intermediate_sets<F: FieldExt, I, Q: Query<F, Eval = F>>(
+    queries: I,
+) -> IntermediateSets<F, Q>
+where
+    I: IntoIterator<Item = Q> + Clone,
+{
+    let queries = queries.into_iter().collect::<Vec<_>>();
+
+    // Find evaluation of a commitment at a rotation
+    let get_eval = |commitment: Q::Commitment, rotation: F| -> F {
+        queries
+            .iter()
+            .find(|query| query.get_commitment() == commitment && query.get_point() == rotation)
+            .unwrap()
+            .get_eval()
+    };
+
+    // Order points according to their rotation
+    let mut rotation_point_map = BTreeMap::new();
+    for query in queries.clone() {
+        let point = rotation_point_map
+            .entry(query.get_point())
+            .or_insert_with(|| query.get_point());
+
+        // Assert rotation point matching consistency
+        assert_eq!(*point, query.get_point());
+    }
+    // All points appear in queries
+    let super_point_set: Vec<F> = rotation_point_map.values().cloned().collect();
+
+    // Collect rotation sets for each commitment
+    // Example elements in the vector:
+    // (C_0, {r_5}),
+    // (C_1, {r_1, r_2, r_3}),
+    // (C_2, {r_2, r_3, r_4}),
+    // (C_3, {r_2, r_3, r_4}),
+    // ...
+    let mut commitment_rotation_set_map: Vec<(Q::Commitment, Vec<F>)> = vec![];
+    for query in queries.clone() {
+        let rotation = query.get_point();
+        if let Some(pos) = commitment_rotation_set_map
+            .iter()
+            .position(|(commitment, _)| *commitment == query.get_commitment())
+        {
+            let (_, rotation_set) = &mut commitment_rotation_set_map[pos];
+            if !rotation_set.contains(&rotation) {
+                rotation_set.push(rotation);
+            }
+        } else {
+            commitment_rotation_set_map.push((query.get_commitment(), vec![rotation]));
+        };
+    }
+
+    // Flatten rotation sets and collect commitments that opens against each commitment set
+    // Example elements in the vector:
+    // {r_5}: [C_0],
+    // {r_1, r_2, r_3} : [C_1]
+    // {r_2, r_3, r_4} : [C_2, C_3],
+    // ...
+    let mut rotation_set_commitment_map = Vec::<(Vec<_>, Vec<Q::Commitment>)>::new();
+    for (commitment, rotation_set) in commitment_rotation_set_map.iter() {
+        if let Some(pos) = rotation_set_commitment_map
+            .iter()
+            .position(|(set, _)| *set == *rotation_set)
+        {
```

Comment:
Sorry that I just realized my approach on this might not be optimal: Comparing vector here will make `[a, b]` and `[b, a]` different rotation set, but they are indeed same. We should compare them by `BTreeSet` like this:

```suggestion
        if let Some(pos) = rotation_set_commitment_map
            .iter()
            .position(|(set, _)| BTreeSet::<F>::from_iter(set.iter().cloned()) == BTreeSet::<F>::from_iter(rotation_set.iter().cloned()))
        {
```